Sources/Core/GraphRequestConnection.swift

```diff
@@ -152,4 +150,134 @@ class GraphRequestConnection: GraphRequestConnecting {
+
+      case .failure(let error):
+        result = .failure(error)
+      }
+    }
+  }
+
+  /**
+   A utility method for converting the fetched data to a Decodable object.
+   This will initially attempt to extract server-side error objects from the response
+
+   - Parameter data: The data used for parsing into valid objects
+   - Parameter remoteType: A generic Decodable type to attempt to parse the data into
+
+   - Returns
+    A Result type with a Success of the generic RemoteType and a Failure of Error
+   */
+  func convertFetchedDataToObjectResult<RemoteType: Decodable>(
+    data: Data,
+    remoteType: RemoteType.Type
+    ) -> Result<RemoteType, Error> {
+    switch try? JSONParser.parse(data: data, for: RemoteGraphResponseError.self) {
+    case let error?:
+      return .failure(error)
+
+    case nil:
+      do {
+        let object = try JSONParser.parse(data: data, for: remoteType)
+        return .success(object)
+      } catch {
+        return .failure(error)
+      }
+    }
+  }
+
+  /**
+   Fetches the data that will later be turned into a Decodable object.
+   This uses a GraphRequest to create a URLRequest, spins up a URLSessionDataTask
+  and calls the completion with either the data from that task or an error
+
+   - Parameter graphRequest: The graph request object to use in creating the data request
+   - Parameter completion: A Result type with a Success of Data and a Failure of Error
+
+   - Returns
+    URLSessionTaskProxy - a wrapper for a url session task that allows you to cancel an in-flight
+   request
+   */
+  func fetchData(
+    for graphRequest: GraphRequest,
+    completion: @escaping (Result<Data, Error>) -> Void
+    ) -> URLSessionTaskProxy? {
+    errorConfiguration = serverConfigurationManager.cachedServerConfiguration?.errorConfiguration ?? errorConfiguration
+
+    piggybackManager.addPiggybackRequests(for: self)
+
+    let urlRequest: URLRequest = self.urlRequest(with: graphRequest)
+
+    logger.log(request: urlRequest, bodyLength: 0, bodyLogger: nil, attachmentLogger: nil)
+
+    requestStartTime = TimeUtility.currentTimeInMilliseconds
+
+    let task = URLSessionTaskProxy(
+      for: urlRequest,
+      fromSession: session
+    ) { data, response, error in
+      let result: Result<Data, Error>
+      defer { completion(result) }
+
+      switch (data, response, error) {
+      case (_, _, let error?):
+        result = .failure(error)
+
+      case (nil, nil, _), (nil, _, nil):
+        result = .failure(GraphRequestConnectionError.missingData)
+
+      case (_, nil, _):
+        result = .failure(GraphRequestConnectionError.missingURLResponse)
+
+      case let (data?, response?, nil):
+        guard let response = response as? HTTPURLResponse else {
+          result = .failure(GraphRequestConnectionError.invalidURLResponseType)
+          return
+        }
+
+        guard response.mimeType?.hasPrefix("image") == false else {
```

Comment:
Since this is an optional, would this be better?
```suggestion
        guard response.mimeType?.hasPrefix("image") != true else {
```